### PR TITLE
Open INI file less

### PR DESCRIPTION
**Pull request by iCollin**
Fixes https://github.com/smartdevicelink/sdl_core/issues/2850

This PR is **ready** for review.

### Risk
This PR makes **no** API changes.

### Testing Plan
Change INI value, run SDL core, ensure Core read new INI value
Unit Tests and ATF tests

### Summary
Refactor INI parsing to reduce the number of times the ini file is opened. This goal is achieved through an update to Profile::UpdateValues:
1. INI file is opened
2. all values from the file are read into a smart_object
3. INI files is closed
4. Member variables are assigned from values in smart_object

##### Enhancements
* SDL Core should open the ini file many times less

##### CMake
* transport_manager now requires smart_object include
* config_profile now depends on smart_object
* config_profile_test now depends on smart_object

### Tasks Remaining:
- [x] Fix unit tests
- [x] Remove ini_file changes
- [x] Deprecate ini_file?

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: src/components/config_profile/src/ini_file.cc [hunk 1]
```diff
@@ -288,131 +280,94 @@ char ini_write_value(const char* fname,
...
-          (temp_str[i] == 10) ||  // LF
-          (temp_str[i] == 13)) {  // CR
-        temp_str[i] = '\0';
-      } else {
-        break;
-      }
-    }
+  const char* equals_ptr = strchr(line_ptr, '=');
+  if (NULL == equals_ptr)
+    return INI_NOTHING;
```
Comment (jacobkeeler) [inline src/components/config_profile/src/ini_file.cc:334]:
```suggestion
  if (NULL == equals_ptr) {
    return INI_NOTHING;
  }
```

File: src/components/config_profile/src/ini_file.cc [hunk 2]
```diff
@@ -288,131 +280,94 @@ char ini_write_value(const char* fname,
...
-        }
-      }
-      return INI_RIGHT_ITEM;
-    } else {
-      return INI_WRONG_ITEM;
+  snprintf(value, INI_LINE_LEN, "%s", temp_str);
+
+  size_t str_len = temp_back - temp_str + 1;
+  for (size_t i = 0; i < str_len; ++i)
+    temp_str[i] = toupper(temp_str[i]);
```
Comment (jacobkeeler) [inline src/components/config_profile/src/ini_file.cc:349]:
```suggestion
  for (size_t i = 0; i < str_len; ++i) {
    temp_str[i] = toupper(temp_str[i]);
  }
```

File: src/components/config_profile/src/ini_file.cc [hunk 3]
```diff
@@ -288,131 +280,94 @@ char ini_write_value(const char* fname,
...
+
+  if (strcmp(temp_str, tag) != 0)
+    return INI_WRONG_ITEM;
```
Comment (jacobkeeler) [inline src/components/config_profile/src/ini_file.cc:352]:
```suggestion
  if (strcmp(temp_str, tag) != 0) {
    return INI_WRONG_ITEM;
  }
```

File: src/components/config_profile/src/ini_file.cc [hunk 4]
```diff
@@ -288,131 +280,94 @@ char ini_write_value(const char* fname,
...
+
+  /* skip leading whitespace on value */
+  line_ptr = equals_ptr + 1;
+  while (is_whitespace(*line_ptr))
+    ++line_ptr;
```
Comment (jacobkeeler) [inline src/components/config_profile/src/ini_file.cc:357]:
```suggestion
  while (is_whitespace(*line_ptr)) {
    ++line_ptr;
  }
```

**Review by jacobkeeler — CHANGES_REQUESTED** [review]

**Comment by iCollin** [comment]
PR description is updated for the new approach

File: src/components/config_profile/src/profile.cc [hunk 5]
```diff
@@ -1253,6 +1223,74 @@ const std::string Profile::hmi_origin_id() const {
...
+      }
+
+      config_obj_[chapter][key] = value;
+    }
+  }
+
+  if (nullptr != config_file_) {
+    fclose(config_file_);
+    config_file_ = nullptr;
+  }
```
Comment (jacobkeeler) [inline src/components/config_profile/src/profile.cc:1292]: This should probably still be it's own function

**Review by jacobkeeler — CHANGES_REQUESTED** [review]
Also, should still deprecate `ini_read_value()`

File: src/components/config_profile/src/profile.cc (hunk 5 above)
Comment (iCollin) [reply src/components/config_profile/src/profile.cc]: Open/Close functions were restored in https://github.com/smartdevicelink/sdl_core/pull/3688/commits/a03c69403fe14790e6b6738a65820c1ae69e95c5

Comment (jacobkeeler) [reply src/components/config_profile/src/profile.cc]: Ah, I was referring to the file parsing code

File: src/components/config_profile/include/config_profile/ini_file.h [hunk 6]
```diff
@@ -87,13 +87,19 @@ extern "C" {
...
  * @brief Write usage instructions to the end of the file
  * @param
- *
+ * 
+ * \deprecated This method will be removed in the next major release
+ * See Profile::UpdateValues for the new ini parsing code
+ * 
  * @return NULL if file or desired entry not found, otherwise pointer to fname
  */
 extern char* ini_write_inst(const char* fname, uint8_t flag);
```
Comment (jacobkeeler) [inline src/components/config_profile/include/config_profile/ini_file.h:96]:
```suggestion
DEPRECATED extern char* ini_write_inst(const char* fname, uint8_t flag);
```

**Review by jacobkeeler — CHANGES_REQUESTED** [review]

File: src/components/config_profile/include/config_profile/ini_file.h (hunk 6 above)
Comment (jacobkeeler) [reply src/components/config_profile/include/config_profile/ini_file.h:96]: Same applies to other deprecated functions

File: src/components/config_profile/src/profile.cc (hunk 5 above)
Comment (iCollin) [reply src/components/config_profile/src/profile.cc:1290]: my bad, https://github.com/smartdevicelink/sdl_core/pull/3688/commits/879fb05d55995b2aeb6054241e509161dedb85c4 was reverted and https://github.com/smartdevicelink/sdl_core/pull/3688/commits/aca5d37da826f7b610bc400060304ad2831fc7f1 moves the ini parsing into its own function

File: src/components/config_profile/include/config_profile/ini_file.h (hunk 6 above)
Comment (iCollin) [reply src/components/config_profile/include/config_profile/ini_file.h]: Ooh I thought this didn't work on C code. Added in https://github.com/smartdevicelink/sdl_core/pull/3688/commits/55c1173c058a08d2fada989db0de86102ae1b990

Also, the ini_file_test.cc file was printing many warnings so I disabled deprecated warnings for that file only in https://github.com/smartdevicelink/sdl_core/pull/3688/commits/59bcec714886c216d7679ff953bef5e94c0741b8, let me know if you don't like that idea. :)

**Review by jacobkeeler — APPROVED** [review]

**Pull request merged**